### PR TITLE
1.Add concat method for Array. 2.Change the declare of  the return value of array push

### PR DESCRIPTION
**Pull request by LiaoPeng**
Signed-off-by: Palmer <pengliao@live.cn>

1.  As a `Array`, concat array is a common requirement.
     So I add a concat method for the class `Array`.
     `concat` method also is the standard interface name of Array.

2. I found the return value of  the `Array push` method is void.
    It should be i32, I think. So I changed it.

### CODE REVIEW

**Comment by MaxGraey** [comment]
Hmm, not sure about non-standard mutable method which can be easily replaced by immutable `Array#concat`. Usually `Array.prototype.push.apply` is not good practice. Could get usage examples?
@dcodeIO wdyt?

**Comment by LiaoPeng** [comment]
@MaxGraey First, I used the name `pushAll`. But the `pushAll` is not the standard name of TypeScript. Consider the code is similar with the TypeScript, I think `concat` maybe better.

**Comment by MaxGraey** [comment]
`Array#concat` method is used to merge two or more arrays. This method does not change the existing arrays, but instead returns a new array. And has current signature:
```ts
 concat(items: T[]): T[];
```

**Comment by LiaoPeng** [comment]
@MaxGraey Thank you for correcting me. 
I have a concern about T[] and Array<T>. 
I found the implements of Array always used `Array<T>`.
But in `index.d.ts` file, the signature declaration  both` T[]` and `Array<T>`  are used.

File: std/assembly/array.ts [hunk 1]
```diff
@@ -153,6 +153,30 @@ export class Array<T> {
     return newLength;
   }
 
+  concat(items: Array<T>): Array<T> {
+    assert(this != null);
+    if (items == null) {
+      items = new Array<T>();
```
Comment (MaxGraey) [inline std/assembly/array.ts]: In JS/TS when you doing `[1,2].concat(null)` result is `[1,2,null]`. But for AS for this case need new conversion. I see two variants whrn items equal to null:
1). Create new array with `len + 1`, copy existing elements to new array and fill last element by zero, so this produce `[1,2,0]` as result.
2). Create new array with `len`, just copy whole existing elements and return `[1,2]` but as cloned.

in both variants we don't need create and link 2 arrays, just only one.

File: std/assembly/index.d.ts [hunk 2]
```diff
@@ -441,7 +441,8 @@ declare class Array<T> {
   includes(searchElement: T, fromIndex?: i32): bool;
   indexOf(searchElement: T, fromIndex?: i32): i32;
   lastIndexOf(searchElement: T, fromIndex?: i32): i32;
-  push(element: T): void;
+  push(items: T): i32;
```
Comment (MaxGraey) [inline std/assembly/index.d.ts]: items -> elements for push arg

File: std/assembly/index.d.ts [hunk 3]
```diff
@@ -441,7 +441,8 @@ declare class Array<T> {
...
+  concat(elements: T[]): T[];
```
Comment (MaxGraey) [inline std/assembly/index.d.ts]: elements -> items for concat arg

File: std/assembly/array.ts (hunk 1 above)
Comment (LiaoPeng) [reply std/assembly/array.ts]: I think the second variant would be better, when the concrete generic type is class.
So I choose the second variant to enhancement.

**Comment by dcodeIO** [comment]
@MaxGraey Do you think this is good to merge?

File: std/assembly/array.ts [hunk 4]
```diff
@@ -153,6 +153,29 @@ export class Array<T> {
...
 
+  concat(items: Array<T>): Array<T> {
+    assert(this != null);
+
+    var thisLen: isize = this.length_;
+    var otherLen = (items == null) ? 0 : items.length_;
+    var outLen =  thisLen + otherLen;
+
+    var out: Array<T> = new Array<T>(outLen);
+    memory.copy(changetype<usize>(out.buffer_) + HEADER_SIZE,
```
Comment (MaxGraey) [inline std/assembly/array.ts]: I think make sense add non-zero length checking also for current array:
```ts
if (thisLen) {
   memory.copy(...)
}
```
And add tests for:
- [].concat(arr)
- arr.concat([])

Comment (LiaoPeng) [reply std/assembly/array.ts]: @MaxGraey I had enhanced the code as you suggested. 
But test case [`assert(source.length == 3);`](https://github.com/LiaoPeng/assemblyscript/blob/d97fb8a21aaac818ab04d128ae7430be4eeab1a8/tests/compiler/std/array.ts#L141) was failed. But I can't find the bugs, can you help me to solve  the issue?

Comment (MaxGraey) [reply std/assembly/array.ts]: Tests should looks like this:
```ts
var source: i32[] = [];
assert(source.length == 0);
out = source.concat(arr);
assert(out.length == 3);
assert(source.length == 0);
```

**Comment by MaxGraey** [comment]
ok, LGTM

File: std/assembly/array.ts [hunk 5]
```diff
@@ -179,6 +179,29 @@ export class Array<T> {
...
+    var otherLen = (items == null) ? 0 : items.length_;
+    var outLen =  thisLen + otherLen;
+    var out: Array<T> = new Array<T>(outLen);
+
+    if (thisLen) {
+      memory.copy(changetype<usize>(out.buffer_) + HEADER_SIZE,
+      changetype<usize>(this.buffer_) + HEADER_SIZE,
+      <usize>(thisLen << alignof<T>()));
+
+      if (isManaged<T>()) __gc_link(changetype<usize>(out), changetype<usize>(this)); // tslint:disable-line
```
Comment (dcodeIO) [inline std/assembly/array.ts]: Why is it registering `this` array as a child (element) of the `out` array here?

Comment (MaxGraey) [reply std/assembly/array.ts]: btw how handle managed references which is array's elements in this case? I also wonder how that would be right?

Comment (MaxGraey) [reply std/assembly/array.ts]: @LiaoPeng remove plz `__gc_link`

Comment (LiaoPeng) [reply std/assembly/array.ts]: @MaxGraey OK.

Comment (MaxGraey) [reply std/assembly/array.ts]: Second one also. @dcodeIO right?

Comment (dcodeIO) [reply std/assembly/array.ts]: Yup

Comment (MaxGraey) [reply std/assembly/array.ts]: Second one also

**Comment by dcodeIO** [comment]
Thank you! :)

**Pull request merged**